crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4231,6 +4232,33 @@ impl<'db> Class<'db> {
 
                     union_of_inferred_types = union_of_inferred_types.add(inferred_ty);
                 }
+                AttributeAssignment::Iterable { iterable } => {
+                    // We found an attribute assignment like:
+                    //
+                    //     for self.name in <iterable>:
+
+                    // TODO: Potential diagnostics resulting from the iterable are currently not reported.
+
+                    let iterable_ty = infer_expression_type(db, *iterable);
+                    let inferred_ty = iterable_ty.iterate(db).unwrap_without_diagnostic();
+
+                    union_of_inferred_types = union_of_inferred_types.add(inferred_ty);
+                }
+                AttributeAssignment::Unpack {
+                    attribute_expression_id,
+                    unpack,
+                } => {
+                    // We found an unpacking assignment like:
+                    //
+                    //     .., self.name, .. = <value>
+                    //     (.., self.name, ..) = <value>
+                    //     [.., self.name, ..] = <value>
+
+                    let inferred_ty = infer_unpack_types(db, *unpack)
+                        .get(*attribute_expression_id)
+                        .expect("Failed to look up type of attribute in unpack assignment");
```

Comment:
I need to think about this again and maybe write a comment here, but I can't currently come up with a scenario in which this would fail.

Reply:
Yes, the `attribute_expression_id` should always be the ID of a target of the unpacking, and unpacking should always populate a type for each target. I guess the latter is not really comprehensively guaranteed at the moment; it depends on the `match target` in `Unpacker::unpack_inner` being comprehensive over all possible target forms. But as long as it covers all the forms that we ever create an assignment for in semantic index building, it will be fine.

Reply:
Yeah, that's correct. The unpacker should populate for each target variable even in the case of length mismatch. I think that wasn't the case in the first iteration but is now guaranteed by this loop:

https://github.com/astral-sh/ruff/blob/a00483a7856cfe5b8c1c59a0e5633a4906d865df/crates/red_knot_python_semantic/src/types/unpacker.rs#L167-L174

Not related to this PR but I think we should also update other places where unpacking is done to consider this a hard error, currently it uses `unwrap_or(Type::unknown())`.

Reply:
We _could_ make this an invariant on Unpacker and update the return type of `get` to `Type` to avoid the `.expect` calls to be spread across similar to how we have methods that directly indexes the inference result.